rust-runtime/aws-smithy-runtime/src/client/runtime_plugin/client_rate_limiter.rs

```diff
@@ -0,0 +1,597 @@
+/// [`StandardRetryStrategy`](crate::client::retries::strategy::standard::StandardRetryStrategy).
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct ClientRateLimiterRuntimePlugin {
+    _rate_limiter: Arc<Mutex<ClientRateLimiter>>,
+}
+
+impl ClientRateLimiterRuntimePlugin {
+    pub fn new(cfg: &ConfigBag) -> Self {
+        Self {
+            _rate_limiter: Arc::new(Mutex::new(ClientRateLimiter::new(cfg))),
+        }
+    }
+}
+
+impl RuntimePlugin for ClientRateLimiterRuntimePlugin {
+    fn config(&self) -> Option<FrozenLayer> {
+        let cfg = Layer::new("client rate limiter");
+        // TODO(enableNewSmithyRuntimeLaunch) Move the Arc/Mutex inside the rate limiter so that it
+        //    be both storable and cloneable.
+        // cfg.store_put(self.rate_limiter.clone());
+
+        Some(cfg.freeze())
+    }
+}
+
+const MIN_FILL_RATE: f64 = 0.5;
+const MIN_CAPACITY: f64 = 1.0;
+const SMOOTH: f64 = 0.8;
+/// How much to scale back after receiving a throttling response
+const BETA: f64 = 0.7;
+/// Controls how aggressively we scale up after being throttled
+const SCALE_CONSTANT: f64 = 0.4;
+
+#[derive(Clone, Debug)]
+pub(crate) struct ClientRateLimiter {
+    /// The rate at which token are replenished.
+    token_refill_rate: f64,
+    /// The maximum capacity allowed in the token bucket.
+    maximum_bucket_capacity: f64,
+    /// The current capacity of the token bucket.
+    /// The minimum this can be is 1.0
+    current_bucket_capacity: f64,
+    /// The last time the token bucket was refilled.
+    time_of_last_refill: Option<f64>,
+    /// The smoothed rate which tokens are being retrieved.
+    tokens_retrieved_per_second: f64,
+    /// The last half second time bucket used.
+    previous_time_bucket: f64,
+    /// The number of requests seen within the current time bucket.
+    request_count: u64,
+    /// Boolean indicating if the token bucket is enabled.
+    /// The token bucket is initially disabled.
+    /// When a throttling error is encountered it is enabled.
+    enable_throttling: bool,
+    /// The maximum rate when the client was last throttled.
+    tokens_retrieved_per_second_at_time_of_last_throttle: f64,
+    /// The last time when the client was throttled.
+    time_of_last_throttle: f64,
+    time_window: f64,
+    calculated_rate: f64,
+}
+
+impl Storable for ClientRateLimiter {
+    type Storer = StoreReplace<Self>;
+}
+
+impl ClientRateLimiter {
+    pub(crate) fn new(cfg: &ConfigBag) -> Self {
+        Self::builder()
+            .time_of_last_throttle(get_unix_timestamp(cfg))
+            .previous_time_bucket(get_unix_timestamp(cfg).floor())
+            .build()
+    }
+
+    fn builder() -> Builder {
+        Builder::new()
+    }
+
+    /// If this function returns `Ok(())`, you're OK to send a request. If it returns an error,
+    /// then you should not send a request; You've sent quite enough already.
+    pub(crate) fn acquire_permission_to_send_a_request(
```

Comment:
Maybe this method takes `timestamp: f64` instead of `cfg: &ConfigBag` since it uses `cfg` only to extract `timestamp`?